### PR TITLE
[opengl] Add import of external OpenGL context.

### PR DESCRIPTION
**Pull request by SamInnopeak**
Issue: #8183

### Brief Summary
Adds the ability to import an external OpenGL context as well as specify if using OpenGLES.

### Walkthrough
* Import OpenGL context added with implementation of ti_import_opengl_runtime()
    * ti_import_opengl_runtime sets if using OpenGLES and the process address used to initialize OpenGL with glad.
    * initialize_opengl will now check to see if there is an existing context before it tries to make a new one.

### <samp>🤖 Generated by Copilot at d340716</samp>

*  Add a `use_gles` parameter to the `ti_import_opengl_runtime` function to allow the user to choose between OpenGL ES and OpenGL when importing an existing OpenGL context ([link](https://github.com/taichi-dev/taichi/pull/8185/files?diff=unified&w=0#diff-147f3f663ebf6052b4fefa43ec1fcdfbc84b63cccb10799e53da38bed9649dfeL35-R36), [link](https://github.com/taichi-dev/taichi/pull/8185/files?diff=unified&w=0#diff-49357769896e790fc2a5882464d92c2eb205ec2473db22480ffe5bee895fea58R21-R30))
*  Define a global variable `kSetGles` to store the user's preference for using OpenGL ES or OpenGL and initialize it to `false` by default ([link](https://github.com/taichi-dev/taichi/pull/8185/files?diff=unified&w=0#diff-bd45dc7fa2776d5568cf0085496a6eeb3ff30876de1a1c6e4b7e4e81798679b6L30-R35), [link](https://github.com/taichi-dev/taichi/pull/8185/files?diff=unified&w=0#diff-d152cbed3aa0abc7bff544a62bf4dac327fa2c002f40100a816d73a52b39318aR28))
*  Modify the `initialize_opengl` function in `taichi/rhi/opengl/opengl_api.cpp` to skip creating or loading GLFW, EGL, or OpenGL functions if `kGetOpenglProcAddr` is set, and to use the existing or current contexts instead ([link](https://github.com/taichi-dev/taichi/pull/8185/files?diff=unified&w=0#diff-bd45dc7fa2776d5568cf0085496a6eeb3ff30876de1a1c6e4b7e4e81798679b6L49-R55), [link](https://github.com/taichi-dev/taichi/pull/8185/files?diff=unified&w=0#diff-bd45dc7fa2776d5568cf0085496a6eeb3ff30876de1a1c6e4b7e4e81798679b6L66-R78), [link](https://github.com/taichi-dev/taichi/pull/8185/files?diff=unified&w=0#diff-bd45dc7fa2776d5568cf0085496a6eeb3ff30876de1a1c6e4b7e4e81798679b6L88-R190))

### CODE REVIEW

File: taichi/rhi/opengl/opengl_api.cpp [hunk 1]
```diff
@@ -27,8 +27,12 @@ static bool kUseGles = false;
 static std::optional<bool> supported;  // std::nullopt
 static bool context_with_glfw = false;
 void *kGetOpenglProcAddr;
+bool kSetGles = false;
```
Comment (bobcao3) [inline taichi/rhi/opengl/opengl_api.cpp:30]: I think GLES "override" might be a better name.

In addition to that, setting a global variable directly doesn't feel too good, might have weird ABI or linking issues. I would prefer this kind of solution with anonymous namespace & std::optional & setter
```c++
namespace {
static std::optional<bool> use_gles_override;
};

static void set_gles_override(bool value) {
  use_gles_override = value;
}

static void unset_gles_override(bool value) {
  use_gles_override = std::nullopt;
}
```

This way later down the code this flag can force GLES both ways

Comment (bobcao3) [reply taichi/rhi/opengl/opengl_api.cpp:30]: And the headers now only need to contain declarations of these two functions

**Review by bobcao3 — COMMENTED** [review]
otherwise LGTM

File: taichi/rhi/opengl/opengl_api.cpp (hunk 1 above)
Comment (SamInnopeak) [reply taichi/rhi/opengl/opengl_api.cpp]: Thank you for the feedback. I've updated the pull request with the changes you suggested. If you have any other feedback or suggestions, that would be appreciated.

**Comment by SamInnopeak** [comment]
I'm currently getting an error when running the tests on the platform "Windows (not sm70)." This is the only platform this error comes up in. This is the error, 
"ERROR tests\python\test_torch_ad.py::test_torch_ad[arch=opengl-2] - RuntimeError: Arch=[<Arch.opengl: 5>] is not supported
ERROR tests\python\test_torch_ad.py::test_torch_ad[arch=gles-4] - RuntimeError: Arch=[<Arch.gles: 11>] is not supported"
This error happens for multiple tests, but is always the same error. The only difference is which test was running. If anyone has any ideas on what might be causing this error, I would greatly appreciate hearing them.

**Comment by bobcao3** [comment]
> I'm currently getting an error when running the tests on the platform "Windows (not sm70)." This is the only platform this error comes up in. This is the error, "ERROR tests\python\test_torch_ad.py::test_torch_ad[arch=opengl-2] - RuntimeError: Arch=[<Arch.opengl: 5>] is not supported ERROR tests\python\test_torch_ad.py::test_torch_ad[arch=gles-4] - RuntimeError: Arch=[<Arch.gles: 11>] is not supported" This error happens for multiple tests, but is always the same error. The only difference is which test was running. If anyone has any ideas on what might be causing this error, I would greatly appreciate hearing them.

That's interesting... I don't think windows support GLES, this might be a case where a test shouldn't be triggered, or somehow GLES is forced on when it can't be done?

**Comment by ailzhang** [comment]
@SamInnopeak I didn't check the logic in the PR carefully enough but from the log it seems the first gles test had some unexpected side effect and failed all the following opengl/gles tests. Maybe this is a good point to get started on debugging? 
image

**Comment by bobcao3** [comment]
> @SamInnopeak I didn't check the logic in the PR carefully enough but from the log it seems the first gles test had some unexpected side effect and failed all the following opengl/gles tests. Maybe this is a good point to get started on debugging? 
> image
> 

I guess ti.reset or something like that did not reset the forced GLES flags. I would say we need to unset that flag once we have created the runtime / done with the runtime?

**Comment by SamInnopeak** [comment]
I fixed the error by removing the part where it called glfwGetCurrentContext(). This seems to have been causing issues on Windows (not sm70.) What seems to have been happening is that glfwGetCurrentContext() was not returning a nullptr after glfwTerminate() was called on that platform specifically. This meant that subsequent tests would try to retrieve the previous test's context which is no longer valid. Either that or glfwTerminate() was not being called between tests at all.

**Review by bobcao3 — APPROVED** [review]
LGTM

**Pull request merged**